### PR TITLE
Adds support for benchmarking CUDA and ROCm functions

### PR DESCRIPTION
**Pull request by kernhanda**
Additionally, fixes an issue with the CUDA loader not being able to run more than one function per process.

### CODE REVIEW

File: hatlib/callable_func.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                self.init_main(args=args)
+                time: float = self.main(args=args)
+            finally:
+                self.cleanup_main(args=args)
+        finally:
+            self.cleanup_runtime()
+
+        return time
+
+    def benchmark(self, warmup_iters, iters, args):
```
Comment (lisaong) [inline hatlib/callable_func.py:21]: -> float

File: hatlib/benchmark_hat_package.py [hunk 2]
```diff
@@ -74,51 +76,70 @@ def get_perf_counter():
...
+            # generate sufficient input sets to overflow the L3 cache, since we don't know the size of the model
+            # we'll make a guess based on the minimum input set size
+            input_sets = generate_input_sets_for_func(func,
+                                                    input_sets_minimum_size_MB,
+                                                    num_additional=10)
 
-        set_size = 0
-        for i in input_sets[0]:
-            set_size += i.size * i.dtype.itemsize
+            set_size = 0
```
Comment (lisaong) [inline hatlib/benchmark_hat_package.py:87]: indent looks like tab?

Comment (lisaong) [reply hatlib/benchmark_hat_package.py:87]: could be a rendering issue on my side

**Review by lisaong — APPROVED** [review]

File: hatlib/cuda_loader.py [hunk 3]
```diff
@@ -119,44 +145,106 @@ def device_args_to_ptr_list(device_args: List):
...
+        self.arg_infos = [ArgInfo(d) for d in hat_arg_descriptions]
+        self.launch_params = func.launch_parameters
+        self.device_mem = None
+        self.ptrs = None
+        self.stream = None
+        self.start_event = None
+        self.stop_event = None
+        self.exec_time = 0.
+
+    def init_runtime(self):
```
Comment (lisaong) [inline hatlib/cuda_loader.py:165]: looks like this is not doing anything for both cuda and hip

File: hatlib/callable_func.py (hunk 1 above)
Comment (kernhanda) [reply hatlib/callable_func.py:21]: done

File: hatlib/cuda_loader.py (hunk 3 above)
Comment (kernhanda) [reply hatlib/cuda_loader.py:165]: I can remove it for now, if you want

Comment (kernhanda) [reply hatlib/cuda_loader.py:165]: submitting the PR. will submit another if we want to keep the interface simpler until needed.

Comment (lisaong) [reply hatlib/cuda_loader.py:165]: sounds good

**Pull request merged**